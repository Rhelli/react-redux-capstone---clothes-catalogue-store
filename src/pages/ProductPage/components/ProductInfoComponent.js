/* eslint-disable react/no-array-index-key */
import React from 'react';
import PropTypes from 'prop-types';
import { genderParse, caseParse } from '../../../utils/textUtils';
import getColorPreview from '../../../utils/colorUtil';

const ProductInfoComponent = props => {
  const {
    id, gender, type, brandName, productName, price,
    color, colorTool, productDetails, productDesc, madeOf,
  } = props;

  return (
    <div>
      <div>
        <span><p>{genderParse(gender)}</p></span>
        <span><p> · </p></span>
        <span><p>{caseParse(type)}</p></span>
      </div>
      <div>
        <h2>{productName}</h2>
        <h3>{brandName}</h3>
      </div>
      <div>
        <div>
          <span><p>{color}</p></span>
          <span><div style={getColorPreview(colorTool)} /></span>
        </div>
        <div>
          <span>
            <label htmlFor="small">
              <input type="radio" id="small" name="size" value="small" />
              S
            </label>
          </span>
          <span>
            <label htmlFor="medium">
              <input type="radio" id="medium" name="size" value="medium" />
              M
            </label>
          </span>
          <span>
            <label htmlFor="large">
              <input type="radio" id="large" name="size" value="large" />
              L
            </label>
          </span>
        </div>
      </div>
      <div>
        <div>
          <div>
            Like
          </div>
          <div>
            <button type="button">{`${price} →`}</button>
          </div>
        </div>
      </div>
      <div>
        <p>{productDesc}</p>
      </div>
      <div>
        <div>
          <ul>
            {
              productDetails.map((detail, index) => (
                <li key={`${id}-${index}`}>{detail}</li>
              ))
            }
          </ul>
        </div>
      </div>
      <div>
        <div>
          <h4>{madeOf}</h4>
        </div>
        <div>
          <h4>{`ID: ${id}`}</h4>
        </div>
      </div>
    </div>
  );
};

ProductInfoComponent.propTypes = {
  gender: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  brandName: PropTypes.string.isRequired,
  id: PropTypes.number.isRequired,
  productName: PropTypes.string.isRequired,
  price: PropTypes.string.isRequired,
  color: PropTypes.string.isRequired,
  colorTool: PropTypes.string.isRequired,
  productDetails: PropTypes.arrayOf(PropTypes.string).isRequired,
  productDesc: PropTypes.string.isRequired,
  madeOf: PropTypes.string.isRequired,
};

export default ProductInfoComponent;
