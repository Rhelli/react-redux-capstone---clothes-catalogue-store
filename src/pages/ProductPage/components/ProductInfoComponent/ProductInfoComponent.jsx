import React from 'react';
import PropTypes from 'prop-types';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faHeart, faShoppingBag } from '@fortawesome/free-solid-svg-icons';
import { genderParse, caseParse } from '../../../../utils';
import getColorPreview from '../../../../utils/colorUtil';
import styles from './ProductInfoComponent.module.scss';

const ProductInfoComponent = props => {
  const {
    id, gender, type, brandName, productName, price,
    color, colorTool, productDetails, productDesc, madeOf,
  } = props;

  return (
    <div className={styles.productInfoContainer}>
      <div className={styles.productInfoSiteMap}>
        <span><p>{genderParse(gender)}</p></span>
        <span><p>&nbsp;›&nbsp;</p></span>
        <span><p>{caseParse(type)}</p></span>
      </div>
      <div className={styles.productInfoTitleBrand}>
        <h2>{productName}</h2>
        <h3>{brandName}</h3>
      </div>
      <div className={styles.productInfoColorSize}>
        <div className={styles.productInfoColorPreview}>
          <p>{color}</p>
          <div className={styles.colorPreviewContainer}>
            <div style={getColorPreview(colorTool)} />
          </div>
        </div>
        <div className={styles.productInfoSizeSelect}>
          <span className={styles.sizeButtonContainer}>
            <label htmlFor="small">
              <input type="radio" id="small" name="size" value="small" />
              <span>S</span>
            </label>
          </span>
          <span className={styles.sizeButtonContainer}>
            <label htmlFor="medium">
              <input type="radio" id="medium" name="size" value="medium" />
              <span className={styles.sizeLabelM}>M</span>
            </label>
          </span>
          <span className={styles.sizeButtonContainer}>
            <label htmlFor="large">
              <input type="radio" id="large" name="size" value="large" />
              <span>L</span>
            </label>
          </span>
        </div>
      </div>
      <div className={styles.productInfoLikeBuy}>
        <button type="button" className={styles.infoLikeButton}>
          <FontAwesomeIcon icon={faHeart} className={styles.likeButtonIcon} />
        </button>
        <div className={styles.productInfoBuyButton}>
          <button type="button">
            {`£${price}`}
            <FontAwesomeIcon className={styles.buyNowIcon} icon={faShoppingBag} />
          </button>
        </div>
      </div>
      <div className={styles.productDescription}>
        <p>{productDesc}</p>
      </div>
      <div className={styles.productDetailsList}>
        <ul>
          {
            productDetails.map((detail, index) => (
              <li key={`${id}-${index + 1}`}>{detail}</li>
            ))
          }
        </ul>
      </div>
      <div className={styles.productExtraInfo}>
        <div className={styles.madeOf}>
          <p>{madeOf}</p>
        </div>
        <div className={styles.productID}>
          <p>{`ID: ${id}`}</p>
        </div>
      </div>
    </div>
  );
};

ProductInfoComponent.propTypes = {
  gender: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  brandName: PropTypes.string.isRequired,
  id: PropTypes.number.isRequired,
  productName: PropTypes.string.isRequired,
  price: PropTypes.string.isRequired,
  color: PropTypes.string.isRequired,
  colorTool: PropTypes.string.isRequired,
  productDetails: PropTypes.arrayOf(PropTypes.string).isRequired,
  productDesc: PropTypes.string.isRequired,
  madeOf: PropTypes.string.isRequired,
};

export default ProductInfoComponent;
