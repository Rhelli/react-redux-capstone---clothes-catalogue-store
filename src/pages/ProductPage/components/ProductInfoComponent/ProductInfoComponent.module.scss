.productInfoContainer {
  width: 100%;
  padding: 20px;
  background-color: #fdf2e9;
  border-top-right-radius: 25px;
  border-top-left-radius: 25px;
  position: relative;
  top: -40px;
  z-index: 2;
}

.productInfoSiteMap {
  display: flex;
  flex-flow: row nowrap;
  font: 400 12px/12px 'Source Sans Pro', monospace;
  color: #808B96;
  padding: 6px 0;

  span {
    padding: 0 1px;
  }
}

.productInfoTitleBrand {
  padding: 6px 0;
  h2 {
    font: 700 32px/32px 'Playfair Display', monospace;
    margin-bottom: 6px;
  }

  h3 {
    font: 400 22px/22px 'Source Sans Pro', monospace;
    color: #5F5B57;
  }
}

.productInfoColorSize {
  display: flex;
  flex-flow: row nowrap;
  align-items: center;
  padding: 8px 0;

  .productInfoColorPreview {
    display: flex;
    align-items: center;
    flex-flow: row nowrap;

    p {
      font: 400 20px/20px 'Source Sans Pro', monospace;
      padding-bottom: 2px;
    }

    .colorPreviewContainer {
      display: flex;
      justify-content: center;
      align-items: center;
      height: 28px;
      width: 28px;
      margin-left: 10px;
      border-radius: 8px;
      background-color: #ffffff;
-webkit-box-shadow: 0px 0px 7px -2px rgba(130,130,130,1);
-moz-box-shadow: 0px 0px 7px -2px rgba(130,130,130,1);
box-shadow: 0px 0px 7px -2px rgba(130,130,130,1);

      div {
        height: 20px;
        width: 20px;
        border-radius: 6px;
      }
    }
  }

  .productInfoSizeSelect {
    margin-left: auto;
    position: relative;

    .sizeButtonContainer {
      input[type="radio"] {
        width: 35px;
        height: 35px;
        border-radius: 8px;
        background-color: white;
        -webkit-appearance: none;
        -moz-appearance: none;
-webkit-box-shadow: inset 0px 0px 9px -3px rgba(207,207,207,1);
-moz-box-shadow: inset 0px 0px 9px -3px rgba(207,207,207,1);
box-shadow: inset 0px 0px 9px -3px rgba(207,207,207,1);
      }

      input[type="radio"]:focus {
        outline: none;
      }

      input[type="radio"]:checked {
        background-color: #D5F5E3;
      }

      input[type="radio"]:checked ~ span:first-of-type {
        font-weight: 700;
      }

      label span {
        position: relative;
        text-align: center;
        left: -26px;
        top: -10px;
        font: 400 22px/12px 'Source Sans Pro', monospace;
        color:  #404040;
      }

      .sizeLabelM {
        left: -28px;
      }


    }
  }
}
