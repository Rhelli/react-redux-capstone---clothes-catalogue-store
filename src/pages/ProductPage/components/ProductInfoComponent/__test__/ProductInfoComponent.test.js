import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import renderer from 'react-test-renderer';
import ProductInfoComponent from '../ProductInfoComponent';

const id = 121111;
const gender = 'male';
const brandName = 'ASOS';
const productName = 'Batman Cape';
const type = 'Cape';
const price = '99';
const color = 'Black';
const colorTool = 'black';
const productDetails = [
  'Makes you the night',
  'Mysterious',
  'Rich and Compelling',
];
const productDesc = 'Something something Dark Knight';
const madeOf = 'Mournful tears';

const productInfoComponent = (
  <ProductInfoComponent
    id={id}
    gender={gender}
    brandName={brandName}
    productName={productName}
    type={type}
    price={price}
    color={color}
    colorTool={colorTool}
    productDetails={productDetails}
    productDesc={productDesc}
    madeOf={madeOf}
  />
);

describe('The appearance and the functionality of the ProductInfoComponent', () => {
  it('Matches the ProductInfoComponent snapshot', () => {
    const tree = renderer.create(productInfoComponent);
    expect(tree).toMatchSnapshot();
  });

  it('Contains a master container with the correct attributes', () => {
    const { container } = render(productInfoComponent);
    expect(container.firstChild.nodeName).toBe('DIV');
    expect(container.firstChild.classList.contains('productInfoContainer')).toBe(true);
  });

  it('Contains seven sub-containers, each for the important clothing information', () => {
    const { container } = render(productInfoComponent);
    expect(container.firstChild.childElementCount).toEqual(7);
  });

  it('Contains a container for the product site map', () => {
    const { container, getByText } = render(productInfoComponent);
    const infoSiteMap = container.firstChild.firstChild;
    expect(infoSiteMap.firstChild.nodeName).toBe('SPAN');
    getByText('Mens');
    getByText('Cape');
  });

  it('Contains a container for the product name and brand', () => {
    const { container, getByText } = render(productInfoComponent);
    const nameBrand = container.firstChild.firstChild.nextSibling;
    expect(nameBrand.firstChild.nodeName).toBe('H2');
    getByText('Batman Cape');
    expect(nameBrand.lastChild.nodeName).toBe('H3');
    getByText('ASOS');
  });

  it('Contains a container for displaying the color and size selection', () => {
    const { getByText, container } = render(productInfoComponent);
    const colorSize = container.firstChild.firstChild.nextSibling.nextSibling;
    expect(colorSize.firstChild.classList.contains('productInfoColorPreview')).toBe(true);
    expect(colorSize.lastChild.classList.contains('productInfoSizeSelect')).toBe(true);
    getByText('Black');
    expect(colorSize.lastChild.childElementCount).toEqual(3);
    getByText('S');
    getByText('M');
    getByText('L');
  });

  it('Contains the container for displaying the like and buy button', () => {
    const { getByText, container } = render(productInfoComponent);
    const likeBuy = container.firstChild.firstChild.nextSibling.nextSibling.nextSibling;
    expect(likeBuy.childElementCount).toBe(2);
    expect(likeBuy.firstChild.classList.contains('infoLikeButton')).toBe(true);
    expect(likeBuy.firstChild.nodeName).toBe('BUTTON');
    expect(likeBuy.lastChild.classList.contains('productInfoBuyButton')).toBe(true);
    expect(likeBuy.lastChild.nodeName).toBe('DIV');
    expect(likeBuy.lastChild.firstChild.nodeName).toBe('BUTTON');
    getByText('Â£99');
  });

  it('Contains the container for displaying the product description', () => {
    const { container, getByText } = render(productInfoComponent);
    const productDesc = container.firstChild.lastChild.previousSibling.previousSibling;
    expect(productDesc.firstChild.nodeName).toBe('P');
    getByText('Something something Dark Knight');
  });

  it('Contains the container for mapping a list of all the product details', () => {
    const { container, getByText } = render(productInfoComponent);
    const productDetails = container.firstChild.lastChild.previousSibling;
    expect(productDetails.firstChild.nodeName).toBe('UL');
    expect(productDetails.firstChild.childElementCount).toBe(3);
    getByText('Makes you the night');
    getByText('Mysterious');
    getByText('Rich and Compelling');
  });

  it('Contains a container for displaying the madeOf info and the id number', () => {
    const { container, getByText } = render(productInfoComponent);
    const productExtraInfo = container.firstChild.lastChild;
    expect(productExtraInfo.childElementCount).toBe(2);
    expect(productExtraInfo.firstChild.classList.contains('madeOf')).toBe(true);
    getByText('Mournful tears');
    expect(productExtraInfo.lastChild.classList.contains('productID')).toBe(true);
    getByText('ID: 121111');
  });
});
