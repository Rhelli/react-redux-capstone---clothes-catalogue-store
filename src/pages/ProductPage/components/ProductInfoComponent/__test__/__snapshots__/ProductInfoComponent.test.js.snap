// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The appearance and the functionality of the ProductInfoComponent Matches the ProductInfoComponent snapshot 1`] = `
<div
  className="productInfoContainer"
>
  <div
    className="productInfoSiteMap"
  >
    <span>
      <p>
        Mens
      </p>
    </span>
    <span>
      <p>
         › 
      </p>
    </span>
    <span>
      <p>
        Cape
      </p>
    </span>
  </div>
  <div
    className="productInfoTitleBrand"
  >
    <h2>
      Batman Cape
    </h2>
    <h3>
      ASOS
    </h3>
  </div>
  <div
    className="productInfoColorSize"
  >
    <div
      className="productInfoColorPreview"
    >
      <p>
        Black
      </p>
      <div
        className="colorPreviewContainer"
      >
        <div
          style={
            Object {
              "backgroundColor": "#101010",
              "borderBottom": "2px solid #202020",
              "color": "#C3C3C3",
            }
          }
        />
      </div>
    </div>
    <div
      className="productInfoSizeSelect"
    >
      <span
        className="sizeButtonContainer"
      >
        <label
          htmlFor="small"
        >
          <input
            id="small"
            name="size"
            type="radio"
            value="small"
          />
          <span>
            S
          </span>
        </label>
      </span>
      <span
        className="sizeButtonContainer"
      >
        <label
          htmlFor="medium"
        >
          <input
            id="medium"
            name="size"
            type="radio"
            value="medium"
          />
          <span
            className="sizeLabelM"
          >
            M
          </span>
        </label>
      </span>
      <span
        className="sizeButtonContainer"
      >
        <label
          htmlFor="large"
        >
          <input
            id="large"
            name="size"
            type="radio"
            value="large"
          />
          <span>
            L
          </span>
        </label>
      </span>
    </div>
  </div>
  <div
    className="productInfoLikeBuy"
  >
    <button
      className="infoLikeButton"
      type="button"
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-heart fa-w-16 likeButtonIcon"
        data-icon="heart"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M462.3 62.6C407.5 15.9 326 24.3 275.7 76.2L256 96.5l-19.7-20.3C186.1 24.3 104.5 15.9 49.7 62.6c-62.8 53.6-66.1 149.8-9.9 207.9l193.5 199.8c12.5 12.9 32.8 12.9 45.3 0l193.5-199.8c56.3-58.1 53-154.3-9.8-207.9z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </button>
    <div
      className="productInfoBuyButton"
    >
      <button
        type="button"
      >
        £99
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-shopping-bag fa-w-14 buyNowIcon"
          data-icon="shopping-bag"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M352 160v-32C352 57.42 294.579 0 224 0 153.42 0 96 57.42 96 128v32H0v272c0 44.183 35.817 80 80 80h288c44.183 0 80-35.817 80-80V160h-96zm-192-32c0-35.29 28.71-64 64-64s64 28.71 64 64v32H160v-32zm160 120c-13.255 0-24-10.745-24-24s10.745-24 24-24 24 10.745 24 24-10.745 24-24 24zm-192 0c-13.255 0-24-10.745-24-24s10.745-24 24-24 24 10.745 24 24-10.745 24-24 24z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </button>
    </div>
  </div>
  <div
    className="productDescription"
  >
    <p>
      Something something Dark Knight
    </p>
  </div>
  <div
    className="productDetailsList"
  >
    <ul>
      <li>
        Makes you the night
      </li>
      <li>
        Mysterious
      </li>
      <li>
        Rich and Compelling
      </li>
    </ul>
  </div>
  <div
    className="productExtraInfo"
  >
    <div
      className="madeOf"
    >
      <p>
        Mournful tears
      </p>
    </div>
    <div
      className="productID"
    >
      <p>
        ID: 121111
      </p>
    </div>
  </div>
</div>
`;
